rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+(tfdbg)](https://www.tensorflow.org/guide/debugger) up-to-date with TF2's
+execution paradigm.
+
+Currently, tfdbg is compatible with only the `tf.Session` API of TensorFlow 1.x.
+However, users of TF2 have raised questions and issues that indicate needs for
+a dedicated debugger, ones that cannot be easily met by a generic Python
+debugger such as pdb.  The most common examples of such needs involve
+finding the source of numeric instability issues like NaNs (e.g., see [GitHub
+issue](https://github.com/tensorflow/tensorflow/issues/26543) and [StackOverflow
+question](https://stackoverflow.com/questions/55823557/converting-darknet53-gives-nan-results-in-tensorflow-2-0))
+
+## Design Proposal
+
+To expand on the aims listed in the Object section, the proposed API will enable
+three key capabilities:
+
+  - **Capability A**: The ability to intercept eagerly-executing TF ops.
+    Specifically, we want a way to register one or more callbacks that are
+    invoked immediately after an `EagerTensor` has been computed through the
+    eager execution of an op. The callback should provide visibility into the
+    input and output `EagerTensor`s of the execution. If so desired, the
+    callback may _override_ the output `EagerTensor`s and thereby transparently
+    affect downstream eager execution.
+  - **Capability B**: The ability to intercept the creation of symbolic ops
+    during function-to-graph conversion in `tf.function`, including the cases
+    assisted by [AutoGraph](https://www.tensorflow.org/guide/autograph). This
+    will form the basis for simulated stepping through lines of
+    `tf.function`-decorated Python functions to assist debugging of graph
+    construction in TF2. Like in Capability A, the callbacks should be able to
+    override the output symbolic tensors of the op in order to affect all
+    downstream graph ops.
+  - **Capability C**: Similar to Capability A above, we want the ability
+    to intercept the runtime execution of FuncGraphs. Note that this
+    requirement could be folded into Capability A, if FuncGraphs are
+    regarded as a special type of op.
+
+Capability B will enable the interception of TF ops executing inside FuncGraphs
+at _runtime_. Although this is not listed explicitly as a desired capability, it
+is critical to runtime debugging in tfdbg v2. Similar to Capability A above, we
+want access to all the intermediate tensor values computed in a FuncGraph.
+
+### Design: Debug Callbacks for Op Instrumentation
+
+The following API meets the requirements listed above. The code below shows the
+API and the detailed signature and semantics of the callbacks that can be passed
+to the API.
+
+```python
+# Exported publicly as: tf.debugging.op_callback()
+def op_callback(callback_fn):
+  """Intercepts op execution and op creation.
+
+  The `callback_fn` will be invoked immediately after any of the three types
+  of events:
+    - The execution of an op under eager mode,
```

Comment:
Perhaps `callback_fn` should also be called if an op fails with a TypeError, e.g. when trying to add a float16 and a float32 tensor. In this case, there would be no outputs, and `attrs` would be partially unspecified.

The reason I mention this is that I am considering implementing a tool to help debug TypeErrors, by detecting where the float16 and float32 tensors originated, and recommending an appropriate place to insert a cast. This is currently impossible, as when a TypeError occurs there is no way of determining the input tensors to the op that failed with a TypeError. I am working on mixed precision, which is using a mix of float16 and float32, so TypeErrors occur fairly frequently.

/CC @alextp @azaks3

Reply:
Interesting use case. IIUC, the API will be able to help you debug that. The callback will be invoked during upstream graph construction. The callback can store the graph-construction history somewhere (e.g., on disk), so you know how the graph is built, op by op. When the program errors out, you can load the history and reason, op by op, why the dtype is wrong.

There is no particular need to call the callback when the error happens. The information is there in the error itself.

How does that sound?

Reply:
The issue is that when the TypeError occurs, you don't know what inputs caused the TypeError, even if you record the graph-construction history from past successful ops. For example, the following programs raises a TypeError

```
a = tf.constant(1., dtype='float16')
b = tf.constant(1., dtype='float32')
c = a + b
```

I could record the graph-construction history, so I could know how `a` and `b` were created. But when the TypeError occurs, I don't know that the inputs that caused the TypeError were `a` and `b`: They could have been some other tensors that were previously created instead. The TypeError itself does not state the inputs, as in Eager mode, tensors do not have names so there would be no way of identifying them.

If this cannot be resolved, it's no big deal. This is a fairly minor use case.

Reply:
@reedwm I see. It seems to me that this issue can be addressed by printing these two extra pieces of info in the error message for dtype error:
1. The name(s) of all input tensors to the op that failed to be created.
2. The name of the graph being constructed (if any).

Reply:
That works in graph mode, but in Eager mode, tensors do not have names.

Still, this doesn't have to be resolved. I could always make the tool require tf.functions to be used.

Reply:
During offline discussion, we agreed that an callback that errors out when an op's execution or construction fails is outside the scope of the design, and so is a "pre-op" callback. Those features make sense and may be added under the tf.debugging.* namespace in the future.